pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/ComputeStep.java

```diff
@@ -188,29 +190,25 @@ private GenericRecord computeFields(
       GenericRecord record,
       Map<org.apache.avro.Schema, org.apache.avro.Schema> schemaCache,
       TransformContext context) {
-    org.apache.avro.Schema avroSchema = record.getSchema();
-
-    Map<String, Object> inferredFields = new HashMap<>();
-    fields
-        .stream()
-        .filter(field -> field.getType() == null)
-        .forEach(
-            field -> inferredFields.put(field.getName(), field.getEvaluator().evaluate(context)));
 
-    List<Schema.Field> computedFields =
-        fields
-            .stream()
-            .map(
-                field ->
-                    createAvroField(
-                        field,
-                        field.getType() != null
-                            ? field.getType()
-                            : getFieldType(inferredFields.get(field.getName()))))
-            .collect(Collectors.toList());
+    // Evaluate computed fields
+    Map<Schema.Field, Object> evaluatedFields = new LinkedHashMap<>(); // preserves the insertion order of keys
+    for (ComputeField field : fields) {
+      Object value = field.getEvaluator().evaluate(context);
+      ComputeFieldType type = field.getType() == null ? getFieldType(value) : field.getType();
+      Schema.Field avroField;
+      if (type == ComputeFieldType.DECIMAL) {
```

Comment:
Can we put this logic in `createAvroField(field, type, value)` ?

Reply:
sure

btw DECIMAL is the first type I encountered that the schema definition itself relies on the value/other metadata hence the special handing (and the lazy evaluation of schema fields after evaluating the values themselves).